### PR TITLE
WFLY-9746 Server does not start with Artemis configured using JGroups TCP stack

### PR DESCRIPTION
**Pull request by pferraro**
https://issues.jboss.org/browse/WFLY-9746
https://issues.jboss.org/browse/WFLY-5459

PR makes the following changes to fix compatibility w/WF11:
* Introduces new schema and model version
    + Add unit tests for new model/schema versions
    + Migrate existing transformer logic to ExtensionTransformerRegistration leveraging ChainedTransformationDescriptionBuilder.
* Restore the jgroups-stack capability reference to refer to a ChannelFactory capability (i.e. WF11 behavior).
* Finally rename jgroups-channel -> jgroups-cluster (WFLY-5459)
* Add new jgroups-channel attribute for broadcast/discovery groups, referencing a CommandDispatcherFactory capability to distinguish between configuration referencing a JGroups stack vs a JGroups channel
* Deprecate the jgroups-stack attribute.
* Support legacy behavior (i.e. when jgroups-stack is specified instead of a jgroups-channel), by fabricating a JGroups channel resource and setting the jgroups-channel attribute appropriately.
* Add the appropriate transformers
* Add docs describing how to configure broadcast/discovery (these did not previously exist).

### CODE REVIEW

**Comment by pferraro** [comment]
@jmesnil Please review.

File: messaging-activemq/pom.xml [hunk 1]
```diff
@@ -217,5 +217,27 @@
             <type>pom</type>
             <scope>test</scope>
         </dependency>
+
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>wildfly-clustering-common</artifactId>
+            <version>${project.version}</version>
```
Comment (jmesnil) [inline messaging-activemq/pom.xml]: is the version needed?

File: messaging-activemq/pom.xml [hunk 2]
```diff
@@ -217,5 +217,27 @@
...
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>wildfly-clustering-jgroups-extension</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>wildfly-clustering-jgroups-extension</artifactId>
+            <version>${project.version}</version>
```
Comment (jmesnil) [inline messaging-activemq/pom.xml]: is the version needed?

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java [hunk 3]
```diff
@@ -68,15 +71,39 @@ private BroadcastGroupAdd() {
...
+        if (operation.hasDefined(JGROUPS_CLUSTER.getName())) {
+            if (operation.hasDefined(JGROUPS_CHANNEL_FACTORY.getName()) && !operation.hasDefined(JGROUPS_CHANNEL.getName())) {
+                // Handle legacy behavior
+                String channel = operation.get(JGROUPS_CLUSTER.getName()).asString();
+                operation.get(JGROUPS_CHANNEL.getName()).set(channel);
+
+                PathAddress channelAddress = context.getCurrentAddress().getParent().getParent().getParent().append(ModelDescriptionConstants.SUBSYSTEM, "jgroups").append("channel", channel);
+                ModelNode addChannelOperation = Util.createAddOperation(channelAddress);
+                addChannelOperation.get("stack").set(operation.get(JGROUPS_CHANNEL_FACTORY.getName()));
+                System.out.println("Executing " + addChannelOperation);
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java]: should be removed

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/CommonAttributes.java [hunk 4]
```diff
@@ -177,15 +177,33 @@
             .setXmlName("param")
             .build();
 
-    SimpleAttributeDefinition JGROUPS_STACK = create("jgroups-stack", ModelType.STRING)
+    SimpleAttributeDefinition JGROUPS_CHANNEL_FACTORY = create("jgroups-stack", ModelType.STRING)
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/CommonAttributes.java]: it should be annotated with `@Deprecated`

**Comment by jmesnil** [comment]
@pferraro I'll finish reviewing the PR tomorrow

Thanks for doing it, it was more than I expected from you. Unfortunately, the code to bump the messaging-activemq subystem XML version and the messaging transformers update was stuck in the Artemis 2.4.x branch (https://github.com/jmesnil/wildfly/commit/5004bc39fc80b498e2b7d0526051b1dbf494beda).

File: messaging-activemq/pom.xml (hunk 1 above)
Comment (pferraro) [reply messaging-activemq/pom.xml]: For test-jars, yes.

File: messaging-activemq/pom.xml (hunk 2 above)
Comment (pferraro) [reply messaging-activemq/pom.xml]: For test-jars, yes.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java (hunk 3 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/BroadcastGroupAdd.java]: oops.  Removing...

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/CommonAttributes.java (hunk 4 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/CommonAttributes.java]: Will do.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java [hunk 5]
```diff
@@ -63,15 +67,38 @@ private DiscoveryGroupAdd() {
         super(DiscoveryGroupDefinition.ATTRIBUTES);
     }
 
+    @Override
+    public void execute(OperationContext context, ModelNode operation) throws OperationFailedException {
+        // Handle jgroups-channel -> jgroups-cluster rename
+        if (!operation.hasDefined(CommonAttributes.JGROUPS_CLUSTER.getName()) && operation.hasDefined(CommonAttributes.LEGACY_JGROUPS_CLUSTER.getName())) {
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java]: Does this code duplicate BroadcastGroupAdd ?
It that's the case, we could put in it in a single place.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java [hunk 6]
```diff
@@ -167,9 +168,10 @@ public static ResourceDescriptionResolver getResourceDescriptionResolver(final b
         return new StandardResourceDescriptionResolver(prefix.toString(), RESOURCE_NAME, MessagingExtension.class.getClassLoader(), true, useUnprefixedChildTypes);
     }
 
+    @Override
     public void initialize(ExtensionContext context) {
         final SubsystemRegistration subsystemRegistration = context.registerSubsystem(SUBSYSTEM_NAME, CURRENT_MODEL_VERSION);
-        subsystemRegistration.registerXMLElementWriter(new MessagingSubsystemParser_2_0());
+        subsystemRegistration.registerXMLElementWriter(MessagingSubsystemParser_3_0::new);
```
Comment (jmesnil) [inline messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java]: Could you create a static instance for the parser and use it both to here and in `initializeParsers`?

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 7]
```diff
@@ -78,8 +78,11 @@ broadcast-group.group-address.deprecated=Use instead a socket-binding to specify
 broadcast-group.group-address=Deprecated. The multicast address to which the data will be broadcast. It is a class D IP address in the range 224.0.0.0 to 239.255.255.255, inclusive. The address 224.0.0.0 is reserved and is not available for use.
 broadcast-group.group-port.deprecated=Use instead a socket-binding to specify the broadcast group's group port.
 broadcast-group.group-port=Deprecated. The UDP port number used for broadcasting.
-broadcast-group.jgroups-channel=The name used by a JGroups channel to join a cluster.
-broadcast-group.jgroups-stack=The name of a stack defined in the org.jboss.as.clustering.jgroups subsystem that is used to form a cluster.
+broadcast-group.jgroups-cluster=The logical cluster name.
+broadcast-group.jgroups-channel=References the name of a JGroups channel.  If undefined, the default channel will be used.
+broadcast-group.jgroups-channel.deprecated=Deprecated. Use jgroups-cluster instead.
+broadcast-group.jgroups-stack=References the name of a JGroups channel factory.
+broadcast-group.jgroups-stack.deprecated=Deprecated. Use jgroups-channel-ref instead.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: I suppose you mean to use `jgroups-channel` instead

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 8]
```diff
@@ -219,8 +222,11 @@ discovery-group.group-address=(Deprecated) Multicast IP address of the group to
 discovery-group.group-port.deprecated=Use instead a socket-binding to specify the discovery group's group port.
 discovery-group.group-port=(Deprecated) UDP port of the multicast group.
 discovery-group.initial-wait-timeout=Period, in ms, to wait for an initial broadcast to give us at least one node in the cluster.
-discovery-group.jgroups-channel=The name used by a JGroups channel to join a cluster.
-discovery-group.jgroups-stack=The name of a stack defined in the org.jboss.as.clustering.jgroups subsystem that is used to form a cluster.
+discovery-group.jgroups-cluster=The logical cluster name.
+discovery-group.jgroups-channel=References the name of a JGroups channel.  If undefined, the default channel will be used.
+discovery-group.jgroups-channel.deprecated=Deprecated. Use jgroups-cluster instead.
+discovery-group.jgroups-stack=References the name of a JGroups channel factory.
+discovery-group.jgroups-stack.deprecated=Deprecated. Use jgroups-channel-ref instead.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: I suppose you mean to use `jgroups-channel` instead

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_3_0.xsd [hunk 9]
```diff
@@ -0,0 +1,904 @@
...
+                    <xs:attribute name="factory-class" type="xs:string" use="required" />
+                </xs:complexType>
+            </xs:element>
+
+            <xs:element name="broadcast-group" minOccurs="0" maxOccurs="unbounded">
+                <xs:complexType>
+                    <xs:attribute name="name" type="xs:string" use="required" />
+
+                    <xs:attribute name="socket-binding" type="xs:string" use="optional" />
+                    <xs:attribute name="jgroups-stack" type="xs:string" use="optional" />
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_3_0.xsd]: could you add a deprecated annotation, please? 

```
<xs:annotation>
  <xs:documentation>
    Deprecated. Use jgroups-channel instead.
  </xs:documentation>
</xs:annotation>
```

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 10]
```diff
@@ -78,8 +78,11 @@ broadcast-group.group-address.deprecated=Use instead a socket-binding to specify
 broadcast-group.group-address=Deprecated. The multicast address to which the data will be broadcast. It is a class D IP address in the range 224.0.0.0 to 239.255.255.255, inclusive. The address 224.0.0.0 is reserved and is not available for use.
 broadcast-group.group-port.deprecated=Use instead a socket-binding to specify the broadcast group's group port.
 broadcast-group.group-port=Deprecated. The UDP port number used for broadcasting.
-broadcast-group.jgroups-channel=The name used by a JGroups channel to join a cluster.
-broadcast-group.jgroups-stack=The name of a stack defined in the org.jboss.as.clustering.jgroups subsystem that is used to form a cluster.
+broadcast-group.jgroups-cluster=The logical cluster name.
+broadcast-group.jgroups-channel=References the name of a JGroups channel.  If undefined, the default channel will be used.
+broadcast-group.jgroups-channel.deprecated=Deprecated. Use jgroups-cluster instead.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: This does not seem correct. `jgroups-channel` is not flagged as deprecated in BroadcastGroupDefinition.

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties [hunk 11]
```diff
@@ -219,8 +222,11 @@ discovery-group.group-address=(Deprecated) Multicast IP address of the group to
 discovery-group.group-port.deprecated=Use instead a socket-binding to specify the discovery group's group port.
 discovery-group.group-port=(Deprecated) UDP port of the multicast group.
 discovery-group.initial-wait-timeout=Period, in ms, to wait for an initial broadcast to give us at least one node in the cluster.
-discovery-group.jgroups-channel=The name used by a JGroups channel to join a cluster.
-discovery-group.jgroups-stack=The name of a stack defined in the org.jboss.as.clustering.jgroups subsystem that is used to form a cluster.
+discovery-group.jgroups-cluster=The logical cluster name.
+discovery-group.jgroups-channel=References the name of a JGroups channel.  If undefined, the default channel will be used.
+discovery-group.jgroups-channel.deprecated=Deprecated. Use jgroups-cluster instead.
```
Comment (jmesnil) [inline messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: This does not seem correct. `jgroups-channel` is not flagged as deprecated in DiscoveryGroupDefinition.

**Review by jmesnil — CHANGES_REQUESTED** [review]

**Comment by jmesnil** [comment]
@pferraro I've requested a few changes to the PR but they are minor and does not affect the running code.

I tested your branch and confirmed that all the following use cases work with it:

* with `<subsystem xmlns="urn:jboss:domain:messaging-activemq:2.0">`
  * OK `<broadcast-group name="bg-group1" jgroups-stack="udp" jgroups-channel="activemq-cluster" broadcast-period="2000" connectors="connector"/>`
  * OK `<broadcast-group name="bg-group1" jgroups-stack="tcp" jgroups-channel="activemq-cluster" broadcast-period="2000" connectors="connector"/>`
* with `<subsystem xmlns="urn:jboss:domain:messaging-activemq:3.0">`
  * OK `<broadcast-group name="bg-group1" jgroups-channel="ee" jgroups-cluster="activemq-cluster" broadcast-period="2000" connectors="connector"/>`
  * OK `<broadcast-group name="bg-group1" jgroups-stack="udp" jgroups-cluster="activemq-cluster" broadcast-period="2000" connectors="connector"/>`

There is one use case that does not work:
* with `<subsystem xmlns="urn:jboss:domain:messaging-activemq:3.0">`
  * FAIL `<broadcast-group name="bg-group1" jgroups-stack="udp" jgroups-cluster="udp" broadcast-period="2000" connectors="connector"/>`

It fails with the error:

```
10:52:29,982 ERROR [org.jboss.as.controller.management-operation] (ServerService Thread Pool -- 15) WFLYCTL0013: Operation ("add") failed - address: ([
    ("subsystem" => "jgroups"),
    ("channel" => "udp")
]) - failure description: "WFLYCTL0436: Cannot register capability 'org.wildfly.clustering.jgroups.channel-factory.udp' at location '[
    (\"subsystem\" => \"jgroups\"),
    (\"channel\" => \"udp\")
]' as it is already registered in context 'global' at location(s) '[[
    (\"subsystem\" => \"jgroups\"),
    (\"stack\" => \"udp\")
]]'"
```

Using the `udp` value for the `jgroups-cluster` attribute does not work. It seems a bad idea anyway since jgroups-cluster is the logical name of the cluster (so `activemq-cluster` is more correct). However I am not sure if that shows an underlying issue with the cap/req...

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java (hunk 5 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/DiscoveryGroupAdd.java]: Just the 1st if block is the same for both add operation handlers, so yes I can consolidate this.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java (hunk 6 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java]: The registerXMLElementWriter(...) method that accepts an XMLElementWriter instance is deprecated.  If this is OK to use now, why is it deprecated?

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties (hunk 7 above)
Comment (pferraro) [reply messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: Correct. Will fix.

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties (hunk 8 above)
Comment (pferraro) [reply messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: Correct. Will fix.

File: messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_3_0.xsd (hunk 9 above)
Comment (pferraro) [reply messaging-activemq/src/main/resources/schema/wildfly-messaging-activemq_3_0.xsd]: Will do.

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties (hunk 10 above)
Comment (pferraro) [reply messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: This deprecation description is used by CommonAttributes.LEGACY_JGROUPS_CLUSTER.  Now that you mention it, I think I can remove this AttributeDefinition and this deprecation description altogether, as it is only used for the PersistentResourceDefinition descriptions for the 1.0/2.0 parser versions.

File: messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties (hunk 11 above)
Comment (pferraro) [reply messaging-activemq/src/main/resources/org/wildfly/extension/messaging/activemq/LocalDescriptions.properties]: This deprecation description is used by CommonAttributes.LEGACY_JGROUPS_CLUSTER.  Now that you mention it, I think I can remove this AttributeDefinition and this deprecation description altogether, as it is only used for the PersistentResourceDefinition descriptions for the 1.0/2.0 parser versions.

**Comment by pferraro** [comment]
A couple points on this configuration:
"<broadcast-group name="bg-group1" jgroups-stack="udp" jgroups-cluster="udp" broadcast-period="2000" connectors="connector"/>"

1. To your point, "udp" is not a realistic cluster name.  Cluster names are use case specific identifiers meant to distinguish themselves from other logical clusters using the same clustering infrastructure.
2. When we introduce capability constraints, there are going to be values for certain attributes that are no longer valid, if they result in the creation of capabilities whose names already exist somewhere else in the configuration. You can many examples of this in the other subsystems where a given capability is supplied by multiple resources.

File: messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java (hunk 6 above)
Comment (pferraro) [reply messaging-activemq/src/main/java/org/wildfly/extension/messaging/activemq/MessagingExtension.java]: Nevermind - I see that the deprecation annotation was removed in wildfly-core master.

**Comment by pferraro** [comment]
@jmesnil Feedback incorporated.

**Review by jmesnil — APPROVED** [review]

**Comment by jmesnil** [comment]
@pferraro thanks paul!

**Comment by jmesnil** [comment]
@kabir fiy, this PR fixes https://issues.jboss.org/browse/WFLY-9746 that is blocker for WildFly 12

**Pull request merged**